src/fast-path.js

```diff
@@ -245,6 +245,21 @@ FPp.needsParens = function(assumeExpressionContext) {
     return true;
   }
 
+  if (
+    (node.type === "BindExpression" ||
+     node.type === "UpdateExpression" ||
+     node.type === "BinaryExpression" ||
+     node.type === "LogicalExpression" ||
+     node.type === "AssignmentExpression" ||
+     node.type === "ConditionalExpression" ||
+     node.type === "SequenceExpression") &&
+    ((parent.type === "ArrowFunctionExpression" && parent.body === node) ||
+    parent.type === "ExpressionStatement") &&
+    startsWithOpenCurlyBrace(node)
+  ) {
+    return true;
+  }
+
   switch (node.type) {
     case "CallExpression":
       if (
```

Comment:
And also `MemberExpression`, for test case

```js
x => ({}().b)
```

This list of node types should pretty much match the list in `startsWithOpenCurlyBrace`. In fact, now that I say that, a better solution occurs to me: replace this entire `if` with

```js
  if (
    ((parent.type === "ArrowFunctionExpression" && parent.body === node) ||
      parent.type === "ExpressionStatement") &&
    startsWithOpenCurlyBrace(node)
  ) {
    return true;
  }
```